### PR TITLE
Release 2022-07-18

### PR DESCRIPTION
**Pull request by Sneagan**
- Improved issuer caching
- Option to stop issuer rotation
- Remove unneeded guard

### CODE REVIEW

**Comment by Sneagan** [comment]
During testing in stage we encountered the following error which persisted for 20 minutes before clearing. This errors was registered as a temporary error at the time of encountering it. This caused lag to remain at a fixed level and once the temporary error cleared the  messages were processed and the lag went to 0. Successful temporary failure test.

```
{"level":"error","error":"ProvisionedThroughputExceededException: The level of configured provisioned throughput for the table was exceeded. Consider increasing your provisioning level with the UpdateTable API.\n\tstatus code: 400, request id: N3STJK47UGBLK4NJTRDLVHCDEFVV4KQNSO5AEMVJF66Q9ASUAAJG","timestamp":"2022-02-11T15:31:01.819209906Z","caller":"/src/kafka/signed_token_redeem_handler.go:199","message":"Request ffd80785-7030-4f10-ab39-b63ec3b438cc: Token redemption failed"}
```

During the 20 minutes the consumer attempted endlessly to retry until the success case. While this is functionally tolerable, it does fill logs and consume resources. A simple, non-exponential backoff was added to the `ProcessingError` and temporary failure handling to decrease the impact of temporary issues.

File: kafka/main.go [hunk 1]
```diff
@@ -185,13 +271,19 @@ func Emit(producer *kafka.Writer, message []byte, logger *zerolog.Logger) error
...
+		logger.Info().Msg("Generating TLSDialer")
 		tlsDialer, _, err := batgo_kafka.TLSDialer()
 		dialer = tlsDialer
 		if err != nil {
 			logger.Error().Msg(fmt.Sprintf("Failed to initialize TLS dialer: %e", err))
 		}
+	} else {
+		logger.Info().Msg("Generating Dialer")
+		dialer = &kafka.Dialer{
+			Timeout:   10 * time.Second,
```
Comment (sadasant) [inline kafka/main.go:284]: Sorry to ask, shouldn’t this be 20 seconds?

Given that the MaxWait is 20 seconds, the dialer might stop 10 seconds before. I’m not sure what would happen afterwards. Would the dialer start dialing again?

Comment (Sneagan) [reply kafka/main.go:284]: [Timeout](https://github.com/segmentio/kafka-go/blob/e1f5945be0b2e606b6826e9e5b0ec0a9ce68bd93/dialer.go#L39) is how long the client will wait for a connection, where [MaxWait](https://github.com/segmentio/kafka-go/blob/382e96d3016e702bce4f51427f8fc7fc62f47b82/reader.go#L414) is how long a connected consumer will wait for a batch to be delivered over that connection. Honestly, both are probably too high and should be lowered but I don't think they'll conflict.

Comment (sadasant) [reply kafka/main.go:284]: Oh ok ok! Thank you 🙏

**Review by amirsaber — APPROVED** [review]
Looks good for prod release. However for merging further down to master we need to have tests for kafka flow, specifically covering `TolerableEquivalence` cases

**Comment by Sneagan** [comment]
@amirsaber To confirm, [this](https://github.com/brave-intl/challenge-bypass-server/issues/104) is required to merge to `master`, not to take this feature to `ads-prod`, correct?

**Comment by amirsaber** [comment]
Correct

File: kafka/main.go [hunk 2]
```diff
@@ -54,105 +69,176 @@ func StartConsumers(providedServer *server.Server, logger *zerolog.Logger) error
...
+			}
+		}
+		close(results)
+		// Iterate over any failures and get the earliest temporary failure offset
+		var temporaryErrors []*utils.ProcessingError
+		for processingError := range results {
+			if processingError.Temporary {
+				continue
+			} else {
+				temporaryErrors = append(temporaryErrors, processingError)
```
Comment (husobee) [inline kafka/main.go:160]: I think this `processingError` will be non-temporary, as we check in the corresponding if for Temporary status?

**Comment by Sneagan** [comment]
This feature is currently deprioritized.

**Pull request closed without merging**